### PR TITLE
remove appimg vs baseosobj from contenttree

### PR DESCRIPTION
**Pull request by deitch**
volumemgr handles all contenttree downloads. There really isn't a difference between one kind of contenttree and the next, and the distinction makes it hard to simplify baseos improvements.

This is a first attempt at eliminating that distinction, having volumemgr listen to contenttreeconfig of any type, and publish plain contenttreestatus, and have the other publishers and subscribers - baseosmgr and zedagent - simply use them as is.

cc @rvs @eriknordmark @zed-rishabh

### CODE REVIEW

**Comment by deitch** [comment]
I have not had the chance to put this through any paces yet. I want CI to do its thing, but more importantly, I want the people cc'ed above to have a first look at it, see what is obviously right or wrong.

**Comment by deitch** [comment]
Does anyone understand this failure? It is in a vendored package that isn't affected by this PR (nor is any other vendored package).

**Review by zed-rishabh — COMMENTED** [review]
@deitch I don't think it will work. Here are my observations:
1. ```volumemgr``` has two publishers (```pubContentTreeStatus```, ```pubBaseOsContentTreeStatus```) and two subscribers (```subContentTreeConfig```, ```subBaseOsContentTreeConfig```) for handling content trees. volumemgr is not using ```pubBaseOsContentTreeStatus``` after your changes.
2. ```baseosmgr``` is subscribing to ```volumemgr``` for all types of content trees that means ```baseosmgr``` will receive ```ContentTreeStatus``` of app images also. How are we planning to prevent that? (I am not sure will it break the existing ```baseosmgr``` workflow)

IMO, we need to move the logic of publishing ```ContentTreeConfig``` from the ```baseosmgr``` to the ```zedagent```. ```baseosmgr``` will publish the ```ContentTreeRefConfig``` to the ```volumemgr```. With this approach, I think we will be able to clean up the types from the configs.

**Comment by deitch** [comment]
> volumemgr has two publishers (pubContentTreeStatus, pubBaseOsContentTreeStatus) and two subscribers (subContentTreeConfig, subBaseOsContentTreeConfig) for handling content trees. volumemgr is not using pubBaseOsContentTreeStatus after your changes

Correct, that was the intent (although I forgot to remove `pubBaseOsContentTreeStatus`, and will fix it). Now it just has one thing to publish, a `ContentTreeStatus`, and one thing for others to subscribe to. No `AgentScope`.

> baseosmgr is subscribing to volumemgr for all types of content trees that means baseosmgr will receive ContentTreeStatus of app images also. How are we planning to prevent that? (I am not sure will it break the existing baseosmgr workflow)

Correct. It receives all, but since it only responds to those for which it cares by UUID, it should ignore the rest. "Should" being the operative word, and I am hoping that you fine folks will see if I got it right.

**Comment by deitch** [comment]
> IMO, we need to move the logic of publishing ContentTreeConfig from the baseosmgr to the zedagent. baseosmgr will publish the ContentTreeRefConfig to the volumemgr. With this approach, I think we will be able to clean up the types from the configs.

100%. This would be a first step. Once we have this in place, then volumemgr is much simplified and ready to handle it.

**Comment by zed-rishabh** [comment]
@deitch Thanks for the clarification. It makes sense to me.
I think we need some kind of flag in the content tree config in API between EVE and controller. It will help the controller differentiating app content trees from the base content tree while displaying the content tree status on the device status page on UI. After these changes, zedagent will receive the base content tree status as well from the volumemgr and it will be reported in the info message to the controller by EVE.

**Review by rvs — APPROVED** [review]
> Does anyone understand this failure? It is in a vendored package that isn't affected by this PR (nor is any other vendored package).

@deitch -- this is just linuxkit intermittent build failure (which reminds me -- we need to probably fix it at some point) for now I just re-kicked the build and it is fine now. Lets see what Eden/Ring0 say.

**Comment by rvs** [comment]
> @deitch Thanks for the clarification. It makes sense to me.
> I think we need some kind of flag in the content tree config in API between EVE and controller. It will help the controller differentiating app content trees from the base content tree while displaying the content tree status on the device status page on UI. After these changes, zedagent will also receive the base content tree status from the volumemgr and it will be reported in the info message to the controller by EVE.

Hey @zed-rishabh -- so you're saying we need some hint to the UI to be able to filer things based on whether it needs to be visible on screen or not? Perhaps the real solution here is that the part of the UI that renders status of baseOS should actually start feeding off of a different source -- mainly regular Volume/ContentTree stats, no?

**Comment by rvs** [comment]
Look like log test in Eden failed -- not sure if this is intermittent or not -- but surely this needs to be take a look at.

**Comment by deitch** [comment]
> @deitch Thanks for the clarification. It makes sense to me.

I am glad it does to someone @zed-rishabh ; I was not totally sure it did! 😁

> I think we need some kind of flag in the content tree config in API between EVE and controller. It will help the controller differentiating app content trees from the base content tree while displaying the content tree status on the device status page on UI. After these changes, zedagent will also receive the base content tree status from the volumemgr and it will be reported in the info message to the controller by EVE

Do we? I think that we need, when looking at a list of apps, to be able to know their status; same for baseos. In other words, going from one of those (app or base) to contenttree. But do we need it the other way? A tree is a tree is a tree?

**Comment by deitch** [comment]
I am not sure of it either. @giggsoff do you know the meaning of the "eden log test" failure in this PR? Is it something intermittent, or did I actually break something?

**Comment by giggsoff** [comment]
Hi, @deitch . It means, that we cannot get log from EVE for our action in the test (ssh into EVE).

What I see on my PC:
```
cat /var/persist/log/reboot-reason.log
Reboot from agent volumemgr[1455] in partition IMGA EVE version 0.0.0-volumemgr-all-contenttree-ae7a0a54-kvm-amd64 at 2020-12-17T09:00:08.347731466Z: fatal: agent volumemgr[1455]: Can not publish /var/run/volumemgr/ContentTreeStatus.sock since it it already used
```

So, I think, the problem in the same `TopicType` [here](https://github.com/deitch/eve/blob/ae7a0a540b708ce688024e36b28382352079606a/pkg/pillar/cmd/volumemgr/volumemgr.go#L222-L225) and [here](https://github.com/deitch/eve/blob/ae7a0a540b708ce688024e36b28382352079606a/pkg/pillar/cmd/volumemgr/volumemgr.go#L183-L186).

**Comment by deitch** [comment]
Ah, the lingering publication that @zed-rishabh found. Thanks for finding it @giggsoff . I am updating now and will run tests again.

**Comment by deitch** [comment]
Updated. Let us see what it does.

**Comment by deitch** [comment]
Anyone else need to review this before @rvs merges it in?

My understanding is that CI has gotten good enough that I don't have to spend real time doing my manual tests?

**Review by lf-edge-eve — APPROVED** [review]
LGTM

**Comment by giggsoff** [comment]
@deitch please take a look at [artifact](https://github.com/lf-edge/eve/suites/1689918149/artifacts/31708798) from integration test. I see `AllowNonFreePort: Unknown ObjType` inside.

**Comment by deitch** [comment]
Thanks @giggsoff ; Adarsh found the same issue. I see what it is, finding a resolution.

**Comment by deitch** [comment]
Short summary. I am awaiting feedback from some others.

Basically, it looks up policy on whether or not to allow `NonFreePort` based on whether it is a `baseos` contenttree or `appimg` contenttree. But the essence of this PR is that there is no difference when it comes to download and verification: contenttree is a contenttree is a contenttree. I would like to just have a single policy setting.

**Comment by deitch** [comment]
Per the suggestions from @zed-rishabh and @rvs , I have updated this to eliminate the distinction between various `objType`s when checking `AllowNonFreePort`. It basically just uses the policy from `"network.allow.wwan.app.download"` for all content tree, both app and baseos.

We probably should rename it to something more universal, but let's do that in a separate PR, since it will affect controller as well.

Ready for review.

File: pkg/pillar/types/downloadertypes.go
Comment (zed-rishabh) [inline pkg/pillar/types/downloadertypes.go:12]: I think the comment can be removed.

File: pkg/pillar/cmd/volumemgr/context.go
Comment (zed-rishabh) [inline pkg/pillar/cmd/volumemgr/context.go:21]: We are not calling the ```subscription``` function for the content tree. Can be removed from here.

File: pkg/pillar/cmd/volumemgr/context.go
Comment (zed-rishabh) [inline pkg/pillar/cmd/volumemgr/context.go:37]: Ditto for ```publication```

**Comment by deitch** [comment]
Looks like it passes now.

**Review by zed-rishabh — COMMENTED** [review]
Some minor nits
We can also remove ```ObjType``` field from the ```ContentTreeConfig``` and ```ContentTreeStatus``` structures defined in ```types/contenttreetypes.go```

File: pkg/pillar/types/downloadertypes.go
Comment (deitch) [reply pkg/pillar/types/downloadertypes.go:12]: Good point. Done.

File: pkg/pillar/cmd/volumemgr/context.go
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/context.go:21]: Good catch. That whole `context.go` file can be gotten rid of.

File: pkg/pillar/cmd/volumemgr/context.go
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/context.go:37]: Ditto.

**Comment by deitch** [comment]
> We can also remove ObjType field from the ContentTreeConfig and ContentTreeStatus structures defined in types/contenttreetypes.go

Good catch there as well. Done.

**Review by zed-rishabh — APPROVED** [review]
LGTM

**Review by rvs — APPROVED** [review]
LGTM

**Comment by deitch** [comment]
Can we get this merged in?

**Comment by rvs** [comment]
Nope -- 'cuz Ring0 is failing again.

**Comment by deitch** [comment]
Moving this forward.

Now, `baseosmgr` successfully transmits the `ContentTreeConfig` to `volumemgr`, which indeed downloads it all and loads it into containerd. It then updates a `ContentTreeStatus`, which `baseosmgr` does find and respond to.

The problem is that when `baseosmgr` receives the status update, it looks for anything to process based not on the UUID, but on the content tree hash.

This creates 2 problems:

1. When `baseosmgr` first creates the config, it often doesn't even know the hash, especially if it is an OCI image, as we latch the hash from volumemgr via downloader. 
2. Even if it does know the hash, that same image could be used in multiple places. E.g. if we have the same image used for baseosmgr and to run in an ECO as a VM. Pulling it down again (or reusing it), thus creating a new `ContentTreeStatus` with different UUID but same hash, should not cause `baseosmgr` to do anything again.

As an example, here is the `BaseOsConfig` created by `zedagent`:

```json
{
  "UUIDandVersion": {
    "UUID": "a18ac2eb-f53f-4ab0-b925-81b0142044e9",
    "Version": "4"
  },
  "BaseOsVersion": "5.17.0-kvm",
  "ContentTreeConfigList": [
    {
      "ContentID": "a18ac2eb-f53f-4ab0-b925-81b0142044e9",
      "DatastoreID": "470d55a8-89da-4bb8-b50f-7dd8737d0eba",
      "RelativeURL": "lfedge/eve:5.17.0-kvm",
      "Format": 8,
      "ContentSha256": "",
      "MaxDownloadSize": 0,
      "GenerationCounter": 0,
      "DisplayName": "lfedge/eve:5.17.0-kvm"
    }
  ],
  "RetryCount": 0,
  "Activate": true
}
```

Notice that the `ContentSha256` is blank.

Now the `ContentTreeConfig` created by `baseosmgr` due to the above config:

```json
{
  "ContentID": "a18ac2eb-f53f-4ab0-b925-81b0142044e9",
  "DatastoreID": "470d55a8-89da-4bb8-b50f-7dd8737d0eba",
  "RelativeURL": "lfedge/eve:5.17.0-kvm",
  "Format": 8,
  "ContentSha256": "",
  "MaxDownloadSize": 0,
  "GenerationCounter": 0,
  "DisplayName": "lfedge/eve:5.17.0-kvm"
}
```

Same UUID as in the baseosconfig, content sha still is blank.

Now here is the `ContentTreeStatus` created by `volumemgr` in response to the above:

```json
{
  "ContentID": "a18ac2eb-f53f-4ab0-b925-81b0142044e9",
  "DatastoreID": "470d55a8-89da-4bb8-b50f-7dd8737d0eba",
  "DatastoreType": "DsContainerRegistry",
  "RelativeURL": "lfedge/eve@sha256:3aa3a0f850103103ff2e60e2323fd38b758477d17daa9d7ea57f3deb4ccbb8ed",
  "Format": 8,
  "ContentSha256": "3aa3a0f850103103ff2e60e2323fd38b758477d17daa9d7ea57f3deb4ccbb8ed",
  "MaxDownloadSize": 0,
  "GenerationCounter": 0,
  "DisplayName": "lfedge/eve:5.17.0-kvm",
  "HasResolverRef": false,
  "State": 108,
  "TotalSize": 283379869,
  "CurrentSize": 283379869,
  "Progress": 100,
  "FileLocation": "",
  "NameIsURL": false,
  "Blobs": [
    "3aa3a0f850103103ff2e60e2323fd38b758477d17daa9d7ea57f3deb4ccbb8ed",
    "99c8e54e2fa6fe8b72986f189c369947a7a3e12d5fe8a85775e9ab50fac56b9e",
    "a0900be7ab37267ff0c841779d0c74c02faf6b29a71342a25337a4ac45ca98c2",
    "38c988dff6ee7b460fba8879eb1c223eee78b8500501d797b4d90b648827af00",
    "7ad2a41413fc25c73feba105bcd1e8a95c43018a3c5e81f8c3578bc7acdc4436",
    "894ecc33944315ceeecf4182ebb799f4e9fa5c37cc08eb20bc3c626cc4ddde4c",
    "1f6f442fb5ebf1c97cbbd445968094192a10567117f4fc57fd8ec0646a513d78",
    "c50d002e168b72acd691b324a6e1de95b0606374d22a3b6932739091ba1466d1",
    "c32fc83ffa1a4234babc50d5fb001dfb65da7bb4d07b678d5772b3b8c5e411d2",
    "b75fa0789eb2395c7774b95a9b68b486a34b4ff899855ee1e50256fa410c4c0e"
  ],
  "ErrorSourceType": "",
  "Error": "",
  "ErrorTime": "0001-01-01T00:00:00Z"
}
```

The content ID is identical (as it should be), and the hash is set (as volumemgr latched it). 

Along comes baseosmgr and tries to match that `ContentTreeStatus` to the source `ContentTreeConfig` using _hash_, and it fails to find it.

I propose that we change baseosmgr to find content based on its UUID. The very response from volumemgr - `ContentTreeStatus` - not only gives the status, but also gives the content sha for that tree.

**Comment by deitch** [comment]
Updated. It now updates based on uuid, rather than hash. Testing.

**Comment by deitch** [comment]
And rebased.

**Comment by deitch** [comment]
Update: `baseosmgr` now correctly catches the update from `volumemgr`. It fails to install, though, because of a tag/name mismatch.

* `8ac34e67-81fd-4bd5-9d11-7be90823be5b-lfedge/eve:5.17.0-kvm` - this is what `baseosmgr` looks for
* `8ac34e67-81fd-4bd5-9d11-7be90823be5b-lfedge/eve@sha256:3aa3a0f850103103ff2e60e2323fd38b758477d17daa9d7ea57f3deb4ccbb8ed` - this is what is in containerd

The first is because it tries to build the reference based on `ContentTreeStatus` that is embedded in `BaseOsStatus`.

Digging through it, I cannot figure out where the right place to apply the latch to `BaseOsStatus` is, and if we do so, how it won't mess up the check that the `RelativeURL` is the same.

**Comment by rvs** [comment]
Hey @deitch -- I see new commits popping up here -- any further comments? Should we proceed with testing this?

**Comment by deitch** [comment]
I am working on it right now @rvs . Some odd issues, but getting there. I will comment here when my local tests are clean.

**Comment by deitch** [comment]
This is at the point where all of the interim steps are in place, with `baseosmgr` doing the right thing.

However, it calls the library to pull the image from containerd expecting a particular format, i.e. the one defined in [edge-containers](https://github.com/lf-edge/edge-containers).

The image we are using for these, e.g. `oci://docker.io/lfedge/eve:5.17.0-kvm`, has a different internal format. We can support it, if we know what it is.

I think @rvs best knows what the layout is inside there. Is it layers with annotations? Or are they just files in certain predefined locations on disk? Let's describe the format here, and then we should be able to support it.

**Review by rvs — APPROVED** [review]
LGTM. Lets test it.

**Comment by rvs** [comment]
> The image we are using for these, e.g. oci://docker.io/lfedge/eve:5.17.0-kvm, has a different internal format. We can support it, if we know what it is.

The high order bit here @deitch is that we need to support a format of a container image where rootfs is just a file inside of the image (and not a special kind of a layer). In fact I'd argue that it would be useful to support it to run user-defined VMs as well.

If we agree to that in principle -- we can discuss the location of the files that need to be extracted and used.

Please let me know what you think @deitch 

> I think @rvs best knows what the layout is inside there. Is it layers with annotations? Or are they just files in certain predefined locations on disk? Let's describe the format here, and then we should be able to support it.

Currently the file we need to extract from the container image is `bits/rootfs.img` however, as mentioned above, we are free to change that pathname inside the container.

That said, all official EVE release images we've published so far under `lfedge/eve:<version>` follow the above convention.

**Comment by deitch** [comment]
Oh this _will_ fail as is. It should work for all non-OCI images (as it did before), but OCI will get most of the way there and then error out. We need to get the format working before that will work too.

**Comment by deitch** [comment]
Looking at the edge-containers [README](https://github.com/lf-edge/edge-containers#differentiation-specification) it explains the format it uses for reading the images and how to find the content.

Basically, it looks at the following in descending order:

1. Special media-type, to indicate what is the kernel or rootfs. This obviously does not happen (at least today) with normal `docker build`.
2. "Normal" docker layer media-types, but have annotations, that indicate which file contains which elements (rootfs, kernel, initrd, etc.)

The image needs to contain some form of flags (i.e. annotations on the manifest) to indicate where to find the various parts.

This means that if the following both are true:

* we want continue to use a regular OCI image, so that anything can read and extract it, not only something compliant with the OCI Artifacts spec
* we want to place the files in those places, e.g. `bits/rootfs.img` for the rootfs

Then we should build those images with the extra annotations, so that eve can know where to find what it needs.

Actually, if it had the annotations, I think it would work already (but won't know until we test)

Where do we build this image? It is pretty easy to add the right annotations to it, whether using libraries or command-line tools.

File: pkg/pillar/types/downloadertypes.go
Comment (eriknordmark) [inline pkg/pillar/types/downloadertypes.go:222]: *If* we are going to to this then we should rename the field to not include "app" in the name, and remove the Base variable.

Comment (deitch) [reply pkg/pillar/types/downloadertypes.go:222]: I agree. I was concerned about everything else upstream that might be affected by it, including both open-source and (especially) commercial controllers. I also was concerned with synchronizing versions and compatibility issues.

I suggest we get this in and working as is, using just the `app` one for both, and then a separate PR that focuses just on eliminating the misnaming of the var.

**Comment by deitch** [comment]
@rvs responded that the image is built using [this Dockerfile](https://github.com/lf-edge/eve/blob/master/pkg/eve/Dockerfile.in) and invoked by [this Makefile](https://github.com/lf-edge/eve/blob/master/Makefile#L449)

File: pkg/pillar/types/downloadertypes.go
Comment (eriknordmark) [reply pkg/pillar/types/downloadertypes.go:222]: @deitch No controller knows about the AllowNonFreeAppImages name; that is local to EVE.
In theory a test might use the configitems
nordmark@bobo:~/master/eden$ grep -r network.allow.wwan.app.download
nordmark@bobo:~/master/eden$ grep -r network.allow.wwan.baseos.download
but the greps come up empty including on the commercial controller side.
So please rename and remove the baseosos including from CONFIG-PROPERTIES.md

**Review by eriknordmark — CHANGES_REQUESTED** [review]
@deitch No controller knows about the AllowNonFreeAppImages name; that is local to EVE.
In theory a test might use the configitems
nordmark@bobo:/master/eden$ grep -r network.allow.wwan.app.download
nordmark@bobo:/master/eden$ grep -r network.allow.wwan.baseos.download
but the greps come up empty including on the commercial controller side.
So please rename and remove the baseosos including from CONFIG-PROPERTIES.md

**Comment by deitch** [comment]
Will do @eriknordmark . As it is this still does not work, as it is blocked by some underlying issues relating to pulling artifacts. I am working it through. I will get the changes that you requested in, but we cannot merge anyways until the underlying piece is resolved.

**Comment by rvs** [comment]
Hey @deitch -- is this getting attention as part of your latest work on baseOS image?

**Comment by deitch** [comment]
All of a piece. It's so tantalizingly close I can taste it.

File: pkg/pillar/cmd/baseosmgr/handlebaseos.go
Comment (eriknordmark) [inline pkg/pillar/cmd/baseosmgr/handlebaseos.go:389]: Can't parse this comment

File: pkg/pillar/types/downloadertypes.go
Comment (eriknordmark) [reply pkg/pillar/types/downloadertypes.go:222]: @deitch any update on this?

File: pkg/pillar/cmd/volumemgr/latchcontenthash.go
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/latchcontenthash.go:117]: Does utils.InsertSha insert a sha even if it is already present or require that the caller only call it is there is no sha? If not, the "maybe" part of the name still makes a ton of sense.

**Review by eriknordmark — COMMENTED** [review]
Some additional comments and nitty concerns.

File: pkg/pillar/cmd/baseosmgr/handlebaseos.go
Comment (deitch) [reply pkg/pillar/cmd/baseosmgr/handlebaseos.go:389]: Wow, well that is some funny typos. I have no idea how we got there. Fixing it now.

File: pkg/pillar/cmd/volumemgr/latchcontenthash.go
Comment (deitch) [reply pkg/pillar/cmd/volumemgr/latchcontenthash.go:117]: It is unchanged. This was just extracting it out to `utils` to make it commonly usable.

I am fine with keeping the old name. Updating.

**Comment by deitch** [comment]
Addressed the two concerns @eriknordmark raised today and pushed out. Now going back to look at the older ones.

File: pkg/pillar/types/downloadertypes.go
Comment (deitch) [reply pkg/pillar/types/downloadertypes.go]: Updated.

**Comment by deitch** [comment]
Updated, addressing all of @eriknordmark 's comments (I hope I got them all correctly).

@rvs let's get this through the full test suite. Once it passes, we can merge it in. The modifications to eve for baseos-in-oci, then, will be very small and straightforward: new version of edge-containers library and a minor change or two.

**Comment by deitch** [comment]
This is strange. The failure is in unit tests of workers, but we didn't change any code in workers (other than fixing a log message typo, and that isn't where it is failing). 🤷‍♂️

**Review by eriknordmark — APPROVED** [review]
LGTM

**Comment by deitch** [comment]
CI all green. Let's merge it in?

**Comment by deitch** [comment]
Thanks @eriknordmark 

Working the last bits for baseos in OCI now. That will be a new (and much simpler) PR

**Pull request merged**
